include/osquery/extensions.h

```diff
@@ -0,0 +1,243 @@
+/*
+ *  Copyright (c) 2014, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#pragma once
+
+#include <osquery/core.h>
+#include <osquery/dispatcher.h>
+#include <osquery/flags.h>
+#include <osquery/registry.h>
+
+#ifdef OSQUERY_THRIFT
+#include "Extension.h"
+#include "ExtensionManager.h"
+#else
+#error "Required -DOSQUERY_THRIFT=/path/to/thrift/gen-cpp"
+#endif
+
+namespace osquery {
+
+DECLARE_string(extensions_socket);
+
+namespace extensions {
+
+/**
+ * @brief Helper struct for managing extenion metadata.
+ *
+ * This structure should match the members of Thrift's InternalExtensionInfo.
+ */
+struct ExtensionInfo {
+  std::string name;
+  std::string version;
+  std::string sdk_version;
+};
+
+/**
+ * @brief The Thrift API server used by an osquery Extension process.
+ *
+ * An extension will load and start a thread to serve the ExtensionHandler
+ * Thrift runloop. This handler is the implementation of the thrift IDL spec.
+ * It implements all the Extension API handlers.
+ *
+ */
+class ExtensionHandler : virtual public ExtensionIf {
+ public:
+  ExtensionHandler() {}
+
+  /// Ping an Extension for status and metrics.
+  void ping(ExtensionStatus& _return);
+
+  /**
+   * @brief The Thrift API used by Registry::call for an extension route.
+   *
+   * @param _return The return response (combo Status and PluginResponse).
+   * @param registry The name of the Extension registry.
+   * @param item The Extension plugin name.
+   * @param request The plugin request.
+   */
+  void call(ExtensionResponse& _return,
```

Comment:
The leading underscore is Thrift-style, we could change it for sure, depends on how Thrifty you want to be, but the generated interface code will have the `_return`s.